modules/publication/ajax/FileUpload.php

```diff
@@ -0,0 +1,877 @@
+            'FROM publication_keyword ' .
+            'WHERE Label=:kw',
+            array('kw' => $kw)
+        );
+        // if it doesn't, add it to keyword table and retrieve ID
+        if (!$kwID) {
+            $kwInsert = array('Label' => $kw);
+            $db->insert(
+                'publication_keyword',
+                $kwInsert
+            );
+            $kwID = $db->getLastInsertId();
+        }
+        // add it pub_kw_rel table
+        // get publication ID
+        $pubKWRelInsert = array(
+                           'PublicationID'        => $pubID,
+                           'PublicationKeywordID' => $kwID,
+                          );
+
+        $db->insert(
+            'publication_keyword_rel',
+            $pubKWRelInsert
+        );
+
+    }
+}
+
+/**
+ * Inserts Variables of Interest for the given publication project
+ *
+ * @param int $pubID publication ID
+ *
+ * @return void
+ */
+function insertVOIs(int $pubID) : void
+{
+
+    if (empty($_POST['voiFields'])) {
+        return;
+    }
+    $db        = Database::singleton();
+    $testNames = $db->pselectColWithIndexKey(
+        'SELECT ID, Test_name FROM test_names',
+        array(),
+        'ID'
+    );
+
+    $paramTypes = $db->pselectColWithIndexKey(
+        'SELECT ParameterTypeID, Name FROM parameter_type',
+        array(),
+        'ParameterTypeID'
+    );
+
+    $voiFields = json_decode($_POST['voiFields']);
+    foreach ($voiFields as $vf) {
+        // search test_names for value
+        if (in_array($vf, $testNames)) {
+            $pubTNRelInsert = array(
+                               'TestNameID'    => array_search($vf, $testNames),
+                               'PublicationID' => $pubID,
+                              );
+            $db->insertIgnore(
+                'publication_test_names_rel',
+                $pubTNRelInsert
+            );
+        } elseif (in_array($vf, $paramTypes)) {
+            $ptID = array_search($vf, $paramTypes);
+            $pubParamTypeRelInsert = array(
+                                      'ParameterTypeID' => $ptID,
+                                      'PublicationID'   => $pubID,
+                                     );
+
+            $db->insertIgnore(
+                'publication_parameter_type_rel',
+                $pubParamTypeRelInsert
+            );
+        } else {
+            throw new LorisException("Unknown variable of interest: $vf");
+        }
+    }
+}
+/**
+ * Deletes all inserted data if there is an exception thrown
+ *
+ * @param int $pubID publication ID
+ *
+ * @return void
+ */
+function cleanup(int $pubID) : void
+{
+    $db    = Database::singleton();
+    $where = array('PublicationID' => $pubID);
+
+    $tables = array(
+               'publication_users_edit_perm_rel',
+               'publication_parameter_type_rel',
+               'publication_test_names_rel',
+               'publication_collaborator_rel',
+               'publication_keyword_rel',
```

Comment:
This could be done with foreign key cascading.

@xlecours also suggests wrapping the script in a transaction and rolling back if an exception is caught instead of using a cleanup function.